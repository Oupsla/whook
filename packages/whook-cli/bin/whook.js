#! /usr/bin/env node

const path = require('path');
const run = require('../dist/index').default;
const PROJECT_SRC =
  process.env.PROJECT_SRC || path.join(process.cwd(), 'dist', 'index');
const { prepareEnvironment } = require(PROJECT_SRC);

run({ prepareEnvironment });
