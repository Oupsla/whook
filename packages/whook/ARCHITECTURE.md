[//]: # ( )
[//]: # (This file is automatically generated by the `jsarch`)
[//]: # (module. Do not change it elsewhere, changes would)
[//]: # (be overriden.)
[//]: # ( )
# Architecture Notes



## Server run

Whook exposes a `runServer` function to programmatically spawn
 its server. It is intended to be reusable and injectable so
 that projects can override the whole `whook` default behavior.

[See in context](./src/index.ts#L73-L77)



## Server preparation

Whook exposes a `prepareServer` function to create its server
 configuration. It takes eventually additional injections that
 would be required at a higher level and a
 [Knifecycle](https://github.com/nfroidure/knifecycle)
 containing the bootstrapped environment and allowing
 to complete and run the server.

[See in context](./src/index.ts#L143-L150)



### Root injections

* We need to inject `httpServer` and `process` to bring life to our
 *  server. We also inject `log` for logging purpose and custom other
 *  injected name that were required upfront.

[See in context](./src/index.ts#L164-L168)



## Server environment

The Whook `prepareEnvironment` function aims to provide the complete
 server environment without effectively planning its run. It allows
 to use that environment for CLI or build purposes. It also
 provides a chance to override some services/constants
 before actually preparing the server.

[See in context](./src/index.ts#L181-L187)



### `PWD` env var

The Whook server heavily rely on the process working directory
 to dynamically load contents. We are making it available to
 the DI system as a constant.

[See in context](./src/index.ts#L198-L202)



### `NODE_ENV` env var

Whook has different behaviors depending on the `NODE_ENV` value
 consider setting it to production before shipping.

[See in context](./src/index.ts#L206-L209)



### `WHOOK_PLUGINS` and `PROJECT_SRC`

Whook need to know where to look up for things like
 commands / handlers etc...

[See in context](./src/index.ts#L213-L216)



### Logging

Whook's default logger write to the NodeJS default console
 except for debugging messages where it uses the `debug`
 module so that you can set the `DEBUG` environment
 variable to `whook` and get debug messages in output.

[See in context](./src/index.ts#L219-L224)



### Initializers

Whook's embed a few default initializers proxied from
 `common-services`, `@whook/http-router` or its own
 `src/services` folder. It can be wrapped or overriden,
 at will, later in project's main file.

[See in context](./src/index.ts#L243-L248)



## Environment service

The `ENV` service add a layer of configuration over just using
 node's `process.env` value. Beware that `PWD` and `NODE_ENV` are
 guaranteed to be the exact same than the injected constants.
 It is up to you to decide upstream if you set them via the
 `process.env.NODE_ENV` and `process.cwd()` values or not.

[See in context](./src/services/ENV.ts#L8-L14)



## Base URL

The `BASE_URL` service is intended to provide a base URL where
 the API can be found at. It can be overriden directly via
 injecting it but it is useful to have a usable URL while
 debugging production environnement.

[See in context](./src/services/BASE_URL.ts#L7-L12)



### Environment isolation

Per default, Whook takes the process environment as is
 but since it could lead to leaks when building for
 AWS Lambda or Google Cloud Functions one can isolate
 the process env when building.

[See in context](./src/services/ENV.ts#L64-L69)



### `.env.NODE_ENV` files

You may need to keep some secrets out of your Git
 history. Whook uses `dotenv` to provide your such
 ability.

[See in context](./src/services/ENV.ts#L77-L81)



## `$autoload` service

The default Whook autoloader provides a simple way to
 load the constants, services and handlers of a Whook
 project automatically from the installed whook plugins.

[See in context](./src/services/_autoload.ts#L38-L42)



### Configuration auto loading

Loading the configuration files is done according to the `NODE_ENV`
 environment variable. It basically requires a configuration hash
 where the keys are Knifecycle constants.

Let's load the configuration files as a convenient way
 to create constants on the fly

[See in context](./src/services/_autoload.ts#L183-L190)



### Wrappers auto loading support

We cannot inject the `WRAPPERS` in the auto loader when
 it is dynamically loaded so giving a second chance here
 for `WRAPPERS` to be set.

[See in context](./src/services/_autoload.ts#L126-L130)



### API auto loading

We cannot inject the `API` in the auto loader since
 it is dynamically loaded so doing this during the auto
 loader initialization.

[See in context](./src/services/_autoload.ts#L110-L114)



### Constants

First of all the autoloader looks for constants in the
 previously loaded `CONFIGS` configurations hash.

[See in context](./src/services/_autoload.ts#L198-L201)



### Handlers map

Here, we build the handlers map needed by the router by injecting every
 handler required by the API.

[See in context](./src/services/_autoload.ts#L213-L216)



### Service/handler loading

Finally, we either require the handler/service module if
 none of the previous strategies applyed.

[See in context](./src/services/_autoload.ts#L243-L246)



#### WhookInitializerMap

Whook exports a `WhookInitializerMap` type to help you ensure yours are valid.

[See in context](./src/services/_autoload.ts#L33-L35)



### Service name mapping

In order to be able to substituate easily a service per another
 one can specify a mapping between a service and its substitution.

[See in context](./src/services/_autoload.ts#L170-L173)



#### WhookServiceMap

Whook exports a `WhookServiceMap` type to help you ensure yours are valid.

[See in context](./src/services/_autoload.ts#L29-L31)



### Initializer path mapping

In order to be able to load a service from a given path map
 one can directly specify a path to use for its resolution.

[See in context](./src/services/_autoload.ts#L276-L279)



## IP detection

If no `HOST` configuration is specified in dependencies nor in ENV,
 this service detects the machine host automagically.

[See in context](./src/services/HOST.ts#L7-L10)



## Port detection

If no `PORT` configuration is specified in dependencies nor in ENV,
 this service detects a free port automagically.

[See in context](./src/services/PORT.ts#L7-L10)



## Project dir

Whook needs to know the directory of the project under
 which he is running. It then uses this service to
 automatically detect it.

[See in context](./src/services/PROJECT_DIR.ts#L7-L12)



## Plugins paths

Whook auto loader can look for initializers in a list of
 plugins defined in the `WHOOK_PLUGINS` constant. This
 service computes the path where those plugins source are
 located allowing one to use services/handlers from it.

[See in context](./src/services/WHOOK_PLUGINS_PATHS.ts#L9-L15)

