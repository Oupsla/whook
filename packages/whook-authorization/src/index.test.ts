import { wrapHandlerWithAuthorization } from '.';
import { handler, ServiceInitializer, HandlerInitializer } from 'knifecycle';
import YError from 'yerror';
import {
  BEARER as BEARER_MECHANISM,
  BASIC as BASIC_MECHANISM,
} from 'http-auth-utils';
import { WhookOperation, WhookHandler } from '@whook/whook';

describe('wrapHandlerWithAuthorization', () => {
  const log = jest.fn();
  const authentication = {
    check: jest.fn(),
  };
  const NOOP_OPERATION: WhookOperation = {
    path: '/path',
    method: 'get',
    operationId: 'noopHandler',
    summary: 'Does nothing.',
    tags: ['system'],
    responses: {
      200: {
        description: 'Sucessfully did nothing!',
      },
    },
  };
  const NOOP_AUTHENTICATED_OPERATION: WhookOperation = {
    ...NOOP_OPERATION,
    security: [
      {},
      {
        bearerAuth: ['user', 'admin'],
      },
    ],
  };
  const NOOP_RESTRICTED_OPERATION: WhookOperation = {
    ...NOOP_OPERATION,
    security: [
      {
        bearerAuth: ['user', 'admin'],
      },
    ],
  };
  const BAD_OPERATION: WhookOperation = {
    ...NOOP_OPERATION,
    security: [
      {
        bearerAuth: [],
      },
    ],
  };

  beforeEach(() => {
    log.mockReset();
    authentication.check.mockReset();
  });

  describe('with unauthenticated endpoints', () => {
    it('should work', async () => {
      const noopHandler = handler(
        async () => ({ status: 200 }),
        'getNoop',
      ) as HandlerInitializer<any, any, any>;
      const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
        noopHandler,
      );
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler({}, NOOP_OPERATION);

      expect({
        response,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    });
  });

  describe('with authenticated but not restricted endpoints', () => {
    it('should work with bearer tokens and good authentication check', async () => {
      authentication.check.mockResolvedValue({
        userId: 1,
        scopes: ['user', 'admin'],
      });
      const noopHandler = handler(
        async () => ({ status: 200 }),
        'getNoop',
      ) as HandlerInitializer<any, any, any>;
      const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
        noopHandler,
      );
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler(
        {
          authorization: 'bearer yolo',
        },
        NOOP_AUTHENTICATED_OPERATION,
      );

      expect({
        response,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    });

    it('should work with Bearer tokens and good authentication check', async () => {
      authentication.check.mockResolvedValue({
        userId: 1,
        scopes: ['user', 'admin'],
      });
      const noopHandler = handler(
        async () => ({ status: 200 }),
        'getNoop',
      ) as HandlerInitializer<any, any, any>;
      const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
        noopHandler,
      );
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler(
        {
          authorization: 'Bearer yolo',
        },
        NOOP_AUTHENTICATED_OPERATION,
      );

      expect({
        response,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    });

    it('should work with access tokens and good authentication check', async () => {
      authentication.check.mockResolvedValue({
        userId: 1,
        scopes: ['user', 'admin'],
      });
      const noopHandler = handler(
        async () => ({ status: 200 }),
        'getNoop',
      ) as HandlerInitializer<any, any, any>;
      const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
        noopHandler,
      );
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler(
        {
          access_token: 'yolo',
        },
        NOOP_AUTHENTICATED_OPERATION,
      );

      expect({
        response,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    });

    it('should work with no authentication at all', async () => {
      authentication.check.mockResolvedValue({
        userId: 1,
        scopes: ['user', 'admin'],
      });
      const noopHandler = handler(
        async () => ({ status: 200 }),
        'getNoop',
      ) as HandlerInitializer<any, any, any>;
      const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
        noopHandler,
      );
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler({}, NOOP_AUTHENTICATED_OPERATION);

      expect({
        response,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    });
  });

  describe('with authenticated and restricted endpoints', () => {
    it('should work with bearer tokens and good authentication check', async () => {
      authentication.check.mockResolvedValue({
        userId: 1,
        scopes: ['user', 'admin'],
      });
      const noopHandler = handler(
        async () => ({ status: 200 }),
        'getNoop',
      ) as HandlerInitializer<any, any, any>;
      const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
        noopHandler,
      );
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler(
        {
          authorization: 'Bearer yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );

      expect({
        response,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    });

    it('should work with access tokens and good authentication check', async () => {
      authentication.check.mockResolvedValue({
        userId: 1,
        scopes: ['user', 'admin'],
      });
      const noopHandler = handler(
        async () => ({ status: 200 }),
        'getNoop',
      ) as HandlerInitializer<any, any, any>;
      const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
        noopHandler,
      );
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler(
        {
          access_token: 'yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );

      expect({
        response,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    });
  });

  it('should fail with no operation definition provided', async () => {
    const noopHandler = handler(
      async () => ({ status: 200 }),
      'getNoop',
    ) as HandlerInitializer<any, any, any>;
    const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
      noopHandler,
    );
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler({
        access_token: 'yolo',
      });
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    }
  });

  it('should fail with a mismatch between user and authenticated one', async () => {
    authentication.check.mockResolvedValue({
      userId: 1,
      scopes: ['user', 'admin'],
    });
    const noopHandler = handler(
      async () => ({ status: 200 }),
      'getNoop',
    ) as HandlerInitializer<any, any, any>;
    const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
      noopHandler,
    );
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          access_token: 'yolo',
          userId: 3,
        },
        NOOP_RESTRICTED_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    }
  });

  it('should fail with bad operation definition provided', async () => {
    const noopHandler = handler(
      async () => ({ status: 200 }),
      'getNoop',
    ) as HandlerInitializer<any, any, any>;
    const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
      noopHandler,
    );
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          access_token: 'yolo',
        },
        BAD_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    }
  });

  it('should fail without right scopes', async () => {
    authentication.check.mockResolvedValue({
      userId: 1,
      scopes: [],
    });
    const noopHandler = handler(
      async () => ({ status: 200 }),
      'getNoop',
    ) as HandlerInitializer<any, any, any>;
    const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
      noopHandler,
    );
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          authorization: 'Bearer yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    }
  });

  it('should fail with unallowed mechanisms', async () => {
    authentication.check.mockResolvedValue({
      userId: 1,
      scopes: [],
    });
    const noopHandler = handler(
      async () => ({ status: 200 }),
      'getNoop',
    ) as HandlerInitializer<any, any, any>;
    const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
      noopHandler,
    );
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      MECHANISMS: [BASIC_MECHANISM, BEARER_MECHANISM],
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          authorization: 'Basic yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    }
  });

  it('should fail with not supported auth', async () => {
    authentication.check.mockRejectedValue(
      new YError('E_UNEXPECTED_TOKEN_CHECK'),
    );

    const noopHandler = handler(
      async () => ({ status: 200 }),
      'getNoop',
    ) as HandlerInitializer<any, any, any>;
    const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
      noopHandler,
    );
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          authorization: 'Whatever yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    }
  });

  it('should fail with no authorization at all for secured endpoints', async () => {
    authentication.check.mockRejectedValue(
      new YError('E_UNEXPECTED_TOKEN_CHECK'),
    );

    const noopHandler = handler(
      async () => ({ status: 200 }),
      'getNoop',
    ) as HandlerInitializer<any, any, any>;
    const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
      noopHandler,
    );
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler({}, NOOP_RESTRICTED_OPERATION);
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    }
  });

  it('should fail with access_token disabled', async () => {
    authentication.check.mockRejectedValue(
      new YError('E_UNEXPECTED_TOKEN_CHECK'),
    );

    const noopHandler = handler(
      async () => ({ status: 200 }),
      'getNoop',
    ) as HandlerInitializer<any, any, any>;
    const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
      noopHandler,
    );
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      DEFAULT_MECHANISM: '',
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          access_token: 'yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    }
  });

  it('should proxy authentication errors', async () => {
    authentication.check.mockRejectedValue(
      new YError('E_UNEXPECTED_TOKEN_CHECK'),
    );

    authentication.check.mockRejectedValue(new YError('E_UNAUTHORIZED'));

    const noopHandler = handler(
      async () => ({ status: 200 }),
      'getNoop',
    ) as HandlerInitializer<any, any, any>;
    const wrappedNoodHandlerWithAuthorization = wrapHandlerWithAuthorization(
      noopHandler,
    );
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          authorization: 'Bearer yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      }).toMatchSnapshot();
    }
  });
});
