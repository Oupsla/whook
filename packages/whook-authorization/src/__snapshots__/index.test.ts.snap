// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapHandlerWithAuthorization should fail with a mismatch between user and authenticated one 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "errorCode": "E_UNAUTHORIZED",
  "errorHeaders": Object {},
  "errorParams": Array [
    1,
    3,
  ],
  "httpCode": 403,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
}
`;

exports[`wrapHandlerWithAuthorization should fail with access_token disabled 1`] = `
Object {
  "authenticationChecks": Array [],
  "errorCode": "E_UNAUTHORIZED",
  "errorHeaders": Object {
    "www-authenticate": "Bearer realm=\\"Auth\\"",
  },
  "errorParams": Array [],
  "httpCode": 401,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
    Array [
      "debug",
      "üîê - No authorization found, locking access!",
    ],
  ],
}
`;

exports[`wrapHandlerWithAuthorization should fail with bad operation definition provided 1`] = `
Object {
  "authenticationChecks": Array [],
  "errorCode": "E_MISCONFIGURATION",
  "errorHeaders": Object {},
  "errorParams": Array [
    "Bearer",
    Array [],
  ],
  "httpCode": 500,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
}
`;

exports[`wrapHandlerWithAuthorization should fail with no authorization at all for secured endpoints 1`] = `
Object {
  "authenticationChecks": Array [],
  "errorCode": "E_UNAUTHORIZED",
  "errorHeaders": Object {
    "www-authenticate": "Bearer realm=\\"Auth\\"",
  },
  "errorParams": Array [],
  "httpCode": 401,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
    Array [
      "debug",
      "üîê - No authorization found, locking access!",
    ],
  ],
}
`;

exports[`wrapHandlerWithAuthorization should fail with no operation definition provided 1`] = `
Object {
  "authenticationChecks": Array [],
  "errorCode": "E_OPERATION_REQUIRED",
  "errorHeaders": Object {},
  "errorParams": Array [],
  "httpCode": 500,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
}
`;

exports[`wrapHandlerWithAuthorization should fail with not supported auth 1`] = `
Object {
  "authenticationChecks": Array [],
  "errorCode": "E_UNKNOWN_AUTH_MECHANISM",
  "errorHeaders": Object {},
  "errorParams": Array [
    "Whatever yolo",
  ],
  "httpCode": 400,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
}
`;

exports[`wrapHandlerWithAuthorization should fail with unallowed mechanisms 1`] = `
Object {
  "authenticationChecks": Array [],
  "errorCode": "E_UNALLOWED_AUTH_MECHANISM",
  "errorHeaders": Object {},
  "errorParams": Array [
    "Basic yolo",
  ],
  "httpCode": 400,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
}
`;

exports[`wrapHandlerWithAuthorization should fail without right scopes 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "errorCode": "E_UNAUTHORIZED",
  "errorHeaders": Object {},
  "errorParams": Array [
    Array [],
    Array [
      "user",
      "admin",
    ],
  ],
  "httpCode": 403,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
}
`;

exports[`wrapHandlerWithAuthorization should proxy authentication errors 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "errorCode": "E_UNAUTHORIZED",
  "errorHeaders": Object {
    "www-authenticate": "Bearer realm=\\"Auth\\"",
  },
  "errorParams": Array [],
  "httpCode": 401,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
}
`;

exports[`wrapHandlerWithAuthorization with authenticated and restricted endpoints should work with access tokens and good authentication check 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "response": Object {
    "headers": Object {
      "X-Authenticated": "{\\"userId\\":1,\\"scopes\\":[\\"user\\",\\"admin\\"]}",
    },
    "status": 200,
  },
}
`;

exports[`wrapHandlerWithAuthorization with authenticated and restricted endpoints should work with bearer tokens and good authentication check 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "response": Object {
    "headers": Object {
      "X-Authenticated": "{\\"userId\\":1,\\"scopes\\":[\\"user\\",\\"admin\\"]}",
    },
    "status": 200,
  },
}
`;

exports[`wrapHandlerWithAuthorization with authenticated but not restricted endpoints should work with Bearer tokens and good authentication check 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "response": Object {
    "headers": Object {
      "X-Authenticated": "{\\"userId\\":1,\\"scopes\\":[\\"user\\",\\"admin\\"]}",
    },
    "status": 200,
  },
}
`;

exports[`wrapHandlerWithAuthorization with authenticated but not restricted endpoints should work with access tokens and good authentication check 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "response": Object {
    "headers": Object {
      "X-Authenticated": "{\\"userId\\":1,\\"scopes\\":[\\"user\\",\\"admin\\"]}",
    },
    "status": 200,
  },
}
`;

exports[`wrapHandlerWithAuthorization with authenticated but not restricted endpoints should work with bearer tokens and good authentication check 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "response": Object {
    "headers": Object {
      "X-Authenticated": "{\\"userId\\":1,\\"scopes\\":[\\"user\\",\\"admin\\"]}",
    },
    "status": 200,
  },
}
`;

exports[`wrapHandlerWithAuthorization with authenticated but not restricted endpoints should work with no authentication at all 1`] = `
Object {
  "authenticationChecks": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
    Array [
      "debug",
      "üîì - Optionally authenticated enpoint detected, letting the call pass through!",
    ],
  ],
  "response": Object {
    "status": 200,
  },
}
`;

exports[`wrapHandlerWithAuthorization with unauthenticated endpoints should work 1`] = `
Object {
  "authenticationChecks": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
    Array [
      "debug",
      "üîì - Public endpoint detected, letting the call pass through!",
    ],
  ],
  "response": Object {
    "status": 200,
  },
}
`;
